File: src/lib/wow/formats/generics.ts
/**
 * Network/file helpers ported from wow.export (src/js/generics.js).
 * Only the headless surface is ported (no DOM/UI helpers).
 */
import fs, { promises as fsp } from 'fs';
import http from 'http';
import https from 'https';
import path from 'path';
import zlib from 'zlib';

import { constants } from '@/lib/wow/formats/constants';
import { write } from '@/lib/wow/log';

import { BufferWrapper } from './buffer';

/** Prefer IPv4 + HTTP/1.1 ALPN for Blizzard TACT/CDN; avoids flaky HTTP/2 + Range on some edges. */
const blizzardHttpsAgent = new https.Agent({ keepAlive: true, family: 4, maxSockets: 64 });
const defaultHttpsAgent = new https.Agent({ keepAlive: true, maxSockets: 64 });
const defaultHttpAgent = new http.Agent({ keepAlive: true, maxSockets: 64 });

/**
 * Async wrapper for fetch() with fallback URL support.
 */
export async function get(url: string | string[]): Promise<Response> {
  const fetchOptions: RequestInit = {
    headers: { 'User-Agent': constants.USER_AGENT },
    redirect: 'follow',
  };

  let index = 1;
  let res: Response | null = null;

  const urlStack = Array.isArray(url) ? [...url] : [url];

  while ((res === null || !res.ok) && urlStack.length > 0) {
    const currentUrl = urlStack.shift()!;
    res = await fetch(currentUrl, fetchOptions);
    write(`get -> [${index++}][${res.status}] ${currentUrl}`);
  }

  return res!;
}

/**
 * Dispatch an async handler for an array of items with a limit to how
 * many can be resolving at once.
 */
export async function queue<T>(items: T[], handler: (item: T) => Promise<unknown>, limit: number): Promise<void> {
  return new Promise((resolve, reject) => {
    let free = limit;
    let complete = -1;
    let index = 0;
    const check = () => {
      complete++;
      free++;

      while (free > 0 && index < items.length) {
        handler(items[index]).then(check).catch(reject);
        index++; free--;
      }

      if (complete === items.length) resolve();
    };

    check();
  });
}

/**
 * Process large arrays in chunks, yielding to the event loop between batches
 * (headless equivalent of wow.export's batchWork redraw yielding).
 */
export async function batchWork<T>(
  name: string,
  items: T[],
  handler: (item: T, index: number) => void,
  batchSize = 1000,
): Promise<void> {
  const totalItems = items.length;
  for (let start = 0; start < totalItems; start += batchSize) {
    const end = Math.min(start + batchSize, totalItems);
    for (let i = start; i < end; i++) handler(items[i], i);

    if (end < totalItems) await new Promise<void>((resolve) => { setImmediate(resolve); });
  }
  write('batchWork "%s" processed %d items', name, totalItems);
}

/** Format a number (bytes) to a displayable file size. */
const doOnceCache = new Map<string, { status: 'pending' | 'complete'; result?: unknown }>();

/** Wrap an async function so it only ever runs once; concurrent callers await the first run. */
export function doOnce<T>(key: string, func: () => Promise<T>): () => Promise<T> {
  return async () => {
    if (!doOnceCache.has(key)) {
      doOnceCache.set(key, { status: 'pending' });
      const result = await func();
      doOnceCache.set(key, { result, status: 'complete' });
      return result;
    }
    while (doOnceCache.get(key)!.status !== 'complete') {
      await new Promise((resolve) => { setTimeout(resolve, 100); });
    }
    return doOnceCache.get(key)!.result as T;
  };
}

/** Clear doOnce memoization so DB/character caches can rebuild after a CASC switch. */
export function resetDoOnceCache(): void {
  doOnceCache.clear();
}

export function filesize(bytes: number): string {
  let size = Number.isNaN(bytes) ? 0 : bytes;
  const units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];

  let unitIndex = Math.floor(Math.log(size) / Math.log(1024));
  if (!Number.isFinite(unitIndex) || unitIndex < 0) unitIndex = 0;

  size = Number((size / 1024 ** unitIndex).toFixed(2));
  return `${size} ${units[unitIndex]}`;
}

/** Ping a URL and measure the response time. */
export async function ping(url: string): Promise<number> {
  const pingStart = Date.now();
  await get(url);
  return Date.now() - pingStart;
}

export async function getJSON(url: string): Promise<unknown> {
  const res = await get(url);
  if (!res.ok) throw new Error(`Unable to request JSON from end-point. HTTP ${res.status} ${res.statusText}`);
  return res.json();
}

/** Read a JSON file from disk, returning null on error. */
export async function readJSON(file: string, ignoreComments = false): Promise<unknown | null> {
  try {
    const raw = await fsp.readFile(file, 'utf8');
    if (ignoreComments) {
      return JSON.parse(raw.split(/\r?\n/).filter((e) => !e.startsWith('//')).join('\n'));
    }
    return JSON.parse(raw);
  } catch (e) {
    return null;
  }
}

function isBlizzardCdnHost(hostname: string): boolean {
  return /\.blizzard\.com$/i.test(hostname) || /\.battle\.net$/i.test(hostname);
}

/**
 * Single GET (follows redirects by delegating to requestData).
 * Blizzard hosts: IPv4 + ALPN http/1.1 + dedicated agent.
 */
function requestDataSingleHop(url: string, partialOfs: number, partialLen: number): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    let parsed: URL;
    try {
      parsed = new URL(url);
    } catch (e) {
      reject(new Error(`Invalid URL: ${url}`));
      return;
    }

    const isHttps = parsed.protocol === 'https:';
    const blizzard = isBlizzardCdnHost(parsed.hostname);
    const headers: Record<string, string> = {
      'User-Agent': constants.USER_AGENT,
    };

    if (partialOfs > -1 && partialLen > -1) headers.Range = `bytes=${partialOfs}-${partialOfs + partialLen - 1}`;

    const options: https.RequestOptions & { ALPNProtocols?: string[] } = {
      hostname: parsed.hostname,
      port: parsed.port || (isHttps ? 443 : 80),
      path: parsed.pathname + parsed.search,
      method: 'GET',
      headers,
      agent: isHttps
        ? (blizzard ? blizzardHttpsAgent : defaultHttpsAgent)
        : defaultHttpAgent,
    };

    if (blizzard) {
      options.family = 4;
      if (isHttps) options.ALPNProtocols = ['http/1.1'];
    }

    const protocol = isHttps ? https : http;
    const req = protocol.request(options, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        const loc = res.headers.location;
        if (!loc) {
          reject(new Error('Redirect without Location header'));
          return;
        }
        write(`Got redirect to ${loc}`);
        res.resume();
        resolve(requestData(new URL(loc, url).href, partialOfs, partialLen));
        return;
      }

      if (!res.statusCode || res.statusCode < 200 || res.statusCode > 302) {
        res.resume();
        reject(new Error(`Status Code: ${res.statusCode}`));
        return;
      }

      const chunks: Buffer[] = [];
      res.on('data', (chunk: Buffer) => chunks.push(chunk));
      res.on('end', () => resolve(Buffer.concat(chunks)));
      res.on('error', reject);
    });

    req.on('error', reject);
    req.setTimeout(600000, () => {
      req.destroy();
      reject(new Error('Request timeout'));
    });
    req.end();
  });
}

function isRetryableDownloadError(err: unknown): boolean {
  const code = (err as NodeJS.ErrnoException)?.code;
  const msg = (err as Error)?.message ?? '';
  if (code === 'ECONNRESET' || code === 'ECONNREFUSED' || code === 'ETIMEDOUT' || code === 'EPIPE') return true;
  if (/socket hang up|timeout/i.test(msg)) return true;
  return false;
}

async function requestData(url: string, partialOfs: number, partialLen: number): Promise<Buffer> {
  const maxAttempts = 4;
  let lastErr: unknown;
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      return await requestDataSingleHop(url, partialOfs, partialLen);
    } catch (err) {
      lastErr = err;
      if (!isRetryableDownloadError(err) || attempt === maxAttempts) throw err;
      const delay = Math.min(400 * attempt * attempt, 8000);
      write(`requestData retry ${attempt}/${maxAttempts} after ${(err as Error).message} (wait ${delay}ms)`);
      await new Promise<void>((r) => { setTimeout(r, delay); });
    }
  }
  throw lastErr;
}

/**
 * Download a file (optionally to a local file).
 * GZIP deflation will be used if headers are set.
 * Data is always returned even if `out` is provided.
 */
export async function downloadFile(
  url: string | string[],
  out?: string,
  partialOfs = -1,
  partialLen = -1,
  deflate = false,
): Promise<BufferWrapper> {
  const urlStack = Array.isArray(url) ? url : [url];

  for (const currentUrl of urlStack) {
    try {
      write(`downloadFile -> ${currentUrl}`);

      let data = await requestData(currentUrl, partialOfs, partialLen);

      if (deflate) {
        data = await new Promise<Buffer>((resolve, reject) => {
          zlib.inflate(data, (err, result) => (err ? reject(err) : resolve(result)));
        });
      }

      const wrapped = new BufferWrapper(data);

      if (out) {
        await createDirectory(path.dirname(out));
        await wrapped.writeToFile(out);
      }

      return wrapped;
    } catch (error) {
      write(`Failed to download from ${currentUrl}: ${(error as Error).message}`);
    }
  }

  throw new Error('All download attempts failed.');
}

/** Create all directories in a given path if they do not exist. */
export async function createDirectory(dir: string): Promise<void> {
  await fsp.mkdir(dir, { recursive: true });
}

/** Wrapper for asynchronously checking if a file exists. */
export async function fileExists(file: string): Promise<boolean> {
  try {
    await fsp.access(file);
    return true;
  } catch (e) {
    return false;
  }
}

/** Read a portion of a file. */
export async function readFile(file: string, offset: number, length: number): Promise<BufferWrapper> {
  const fd = await fsp.open(file);
  const buf = BufferWrapper.alloc(length);

  await fd.read(buf.raw, 0, length, offset);
  await fd.close();

  return buf;
}

/** Recursively delete a directory and everything inside of it. Returns total size deleted. */
export async function deleteDirectory(dir: string): Promise<number> {
  let deleteSize = 0;
  try {
    const entries = await fsp.readdir(dir);
    for (const entry of entries) {
      const entryPath = path.join(dir, entry);
      const entryStat = await fsp.stat(entryPath);

      if (entryStat.isDirectory()) {
        deleteSize += await deleteDirectory(entryPath);
      } else {
        await fsp.unlink(entryPath);
        deleteSize += entryStat.size;
      }
    }

    await fsp.rmdir(dir);
  } catch (e) {
    // Something failed to delete.
  }

  return deleteSize;
}

/** Calculate the hash of a file. */
export async function getFileHash(file: string, method: string, encoding: 'hex' | 'base64'): Promise<string> {
  const { createHash } = await import('crypto');
  return new Promise((resolve) => {
    const fd = fs.createReadStream(file);
    const hash = createHash(method);

    fd.on('data', (chunk) => hash.update(chunk));
    fd.on('end', () => resolve(hash.digest(encoding)));
  });
}

File: golang/internal/wow/formats/generics.go
// Package formats provides network and file helpers ported from generics.ts.
package formats

import (
	"bytes"
	"compress/zlib"
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/pqhuy98/wow-converter/internal/buffer"
	"github.com/pqhuy98/wow-converter/internal/wow/constants"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
)

var blizzardHostPattern = regexp.MustCompile(`(?i)\.(blizzard\.com|battle\.net)$`)

// HTTPResponse mirrors fetch Response fields used by the TS port.
type HTTPResponse struct {
	Status int
	Body   []byte
	OK     bool
}

type downloadStatusError struct {
	status int
}

func (e downloadStatusError) Error() string {
	return fmt.Sprintf("status code: %d", e.status)
}

var httpClient = &http.Client{
	Timeout: 10 * time.Minute,
	Transport: &http.Transport{
		MaxIdleConns:        64,
		MaxIdleConnsPerHost: 64,
		IdleConnTimeout:     90 * time.Second,
		TLSClientConfig:     &tls.Config{},
	},
}

// Get performs HTTP GET with fallback URL support.
func Get(urls ...string) (*HTTPResponse, error) {
	if len(urls) == 0 {
		return nil, errors.New("no URL provided")
	}
	index := 1
	var last *HTTPResponse
	for len(urls) > 0 {
		currentURL := urls[0]
		urls = urls[1:]
		resp, err := doGet(currentURL, -1, -1)
		if err != nil {
			return nil, err
		}
		log.Write("get -> [%d][%d] %s", index, resp.Status, currentURL)
		index++
		last = resp
		if resp.OK {
			return resp, nil
		}
	}
	return last, nil
}

func doGet(url string, partialOfs, partialLen int) (*HTTPResponse, error) {
	data, status, err := requestData(url, partialOfs, partialLen)
	if err != nil {
		return nil, err
	}
	return &HTTPResponse{
		Status: status,
		Body:   data,
		OK:     status >= 200 && status <= 299,
	}, nil
}

func isRetryableDownloadError(err error) bool {
	if err == nil {
		return false
	}
	var statusErr downloadStatusError
	if errors.As(err, &statusErr) {
		return statusErr.status == http.StatusRequestTimeout ||
			statusErr.status == http.StatusTooManyRequests ||
			statusErr.status >= http.StatusInternalServerError
	}
	msg := err.Error()
	return strings.Contains(msg, "timeout") ||
		strings.Contains(msg, "connection reset") ||
		strings.Contains(msg, "connection refused") ||
		strings.Contains(msg, "EOF")
}

func requestData(url string, partialOfs, partialLen int) ([]byte, int, error) {
	maxAttempts := 4
	var lastErr error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		data, status, err := requestDataSingleHop(url, partialOfs, partialLen)
		if err == nil {
			return data, status, nil
		}
		lastErr = err
		if !isRetryableDownloadError(err) || attempt == maxAttempts {
			return nil, 0, err
		}
		delay := min(400*attempt*attempt, 8000)
		log.Write("requestData retry %d/%d after %s (wait %dms)", attempt, maxAttempts, err.Error(), delay)
		time.Sleep(time.Duration(delay) * time.Millisecond)
	}
	return nil, 0, lastErr
}

func requestDataSingleHop(url string, partialOfs, partialLen int) ([]byte, int, error) {
	if err := ValidateFetchURL(url); err != nil {
		return nil, 0, err
	}
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, 0, err
	}
	req.Header.Set("User-Agent", constants.UserAgent)
	if partialOfs > -1 && partialLen > -1 {
		req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", partialOfs, partialOfs+partialLen-1))
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusMovedPermanently || resp.StatusCode == http.StatusFound {
		loc := resp.Header.Get("Location")
		if loc == "" {
			return nil, 0, errors.New("redirect without Location header")
		}
		log.Write("Got redirect to %s", loc)
		return requestData(resolveURL(url, loc), partialOfs, partialLen)
	}

	if resp.StatusCode < 200 || resp.StatusCode > 302 {
		return nil, resp.StatusCode, downloadStatusError{status: resp.StatusCode}
	}

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, resp.StatusCode, err
	}
	return data, resp.StatusCode, nil
}

func resolveURL(base, loc string) string {
	if strings.HasPrefix(loc, "http://") || strings.HasPrefix(loc, "https://") {
		return loc
	}
	if strings.HasPrefix(loc, "/") {
		// best-effort join
		if i := strings.Index(base, "://"); i >= 0 {
			rest := base[i+3:]
			if j := strings.Index(rest, "/"); j >= 0 {
				return base[:i+3+j] + loc
			}
		}
	}
	return loc
}

// boundedConcurrency caps a worker limit to the number of jobs (min 1 when jobs > 0).
func boundedConcurrency(limit, jobs int) int {
	if jobs < 1 {
		return 1
	}
	if limit < 1 {
		limit = 1
	}
	if limit > jobs {
		return jobs
	}
	return limit
}

// Queue dispatches handlers with a concurrency limit.
func Queue[T any](items []T, handler func(T) error, limit int) error {
	limit = boundedConcurrency(limit, len(items))
	sem := make(chan struct{}, limit)
	errCh := make(chan error, 1)
	var wg sync.WaitGroup
	for _, item := range items {
		wg.Add(1)
		go func(it T) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()
			if err := handler(it); err != nil {
				select {
				case errCh <- err:
				default:
				}
			}
		}(item)
	}
	wg.Wait()
	select {
	case err := <-errCh:
		return err
	default:
		return nil
	}
}

// BatchWork processes items in chunks, yielding between batches.
func BatchWork[T any](name string, items []T, handler func(T, int), batchSize int) {
	if batchSize <= 0 {
		batchSize = 1000
	}
	total := len(items)
	for start := 0; start < total; start += batchSize {
		end := start + batchSize
		if end > total {
			end = total
		}
		for i := start; i < end; i++ {
			handler(items[i], i)
		}
		if end < total {
			time.Sleep(0)
		}
	}
	log.Write("batchWork \"%s\" processed %d items", name, total)
}

func Filesize(bytes int) string {
	size := float64(bytes)
	if math.IsNaN(size) {
		size = 0
	}
	units := []string{"B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"}
	unitIndex := 0
	if size > 0 {
		unitIndex = int(math.Log(size) / math.Log(1024))
	}
	if unitIndex < 0 {
		unitIndex = 0
	}
	if unitIndex >= len(units) {
		unitIndex = len(units) - 1
	}
	if size > 0 {
		size = size / math.Pow(1024, float64(unitIndex))
	}
	return fmt.Sprintf("%.2f %s", size, units[unitIndex])
}

// Ping measures response time for a URL.
// CDN host resolution pings CDN roots (e.g. https://level3.blizzard.com/) which
// often return 403/404; only transport failures should count as errors (wow.export parity).
func Ping(url string) (int, error) {
	start := time.Now()
	if err := pingRequest(url); err != nil {
		return 0, err
	}
	return int(time.Since(start).Milliseconds()), nil
}

func pingRequest(url string) error {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", constants.UserAgent)
	resp, err := httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusMovedPermanently || resp.StatusCode == http.StatusFound {
		loc := resp.Header.Get("Location")
		if loc == "" {
			return errors.New("redirect without Location header")
		}
		io.Copy(io.Discard, resp.Body)
		return pingRequest(resolveURL(url, loc))
	}

	_, err = io.Copy(io.Discard, resp.Body)
	return err
}

// ReadJSON reads a JSON file, returning nil on error.
func ReadJSON(file string, _ bool) (map[string]string, error) {
	data, err := os.ReadFile(file)
	if err != nil {
		return nil, nil
	}
	var out map[string]string
	if err := jsonUnmarshal(data, &out); err != nil {
		return nil, nil
	}
	return out, nil
}

func jsonUnmarshal(data []byte, out any) error {
	return json.Unmarshal(data, out)
}

// DownloadFile downloads a file with optional partial range and output path.
func DownloadFile(urls []string, out string, partialOfs, partialLen int, deflate bool) (*buffer.Buffer, error) {
	for _, currentURL := range urls {
		log.Write("downloadFile -> %s", currentURL)
		data, _, err := requestData(currentURL, partialOfs, partialLen)
		if err != nil {
			log.Write("Failed to download from %s: %s", currentURL, err.Error())
			continue
		}
		if deflate {
			r, zerr := zlib.NewReader(bytes.NewReader(data))
			if zerr != nil {
				log.Write("Failed to inflate from %s: %s", currentURL, zerr.Error())
				continue
			}
			inflated, rerr := io.ReadAll(r)
			r.Close()
			if rerr != nil {
				log.Write("Failed to read inflated data from %s: %s", currentURL, rerr.Error())
				continue
			}
			data = inflated
		}
		wrapped := buffer.From(data)
		if out != "" {
			if err := CreateDirectory(filepath.Dir(out)); err != nil {
				return nil, err
			}
			if err := wrapped.WriteToFile(out); err != nil {
				return nil, err
			}
		}
		return wrapped, nil
	}
	return nil, errors.New("all download attempts failed")
}

// CreateDirectory creates all directories in a path.
func CreateDirectory(dir string) error {
	return os.MkdirAll(dir, 0o755)
}

// FileExists checks whether a file exists.
func FileExists(file string) bool {
	_, err := os.Stat(file)
	return err == nil
}

// ReadFile reads a portion of a file.
func ReadFile(file string, offset, length int) (*buffer.Buffer, error) {
	f, err := os.Open(file)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	buf := buffer.Alloc(length, false)
	_, err = f.ReadAt(buf.Raw(), int64(offset))
	if err != nil && !errors.Is(err, io.EOF) {
		return nil, err
	}
	return buf, nil
}

// DeleteDirectory recursively deletes a directory.
func DeleteDirectory(dir string) (int, error) {
	var deleteSize int
	entries, err := os.ReadDir(dir)
	if err != nil {
		return 0, nil
	}
	for _, entry := range entries {
		entryPath := filepath.Join(dir, entry.Name())
		if entry.IsDir() {
			s, _ := DeleteDirectory(entryPath)
			deleteSize += s
		} else {
			info, err := entry.Info()
			if err == nil {
				deleteSize += int(info.Size())
			}
			_ = os.Remove(entryPath)
		}
	}
	_ = os.Remove(dir)
	return deleteSize, nil
}

func isBlizzardCdnHost(hostname string) bool {
	return blizzardHostPattern.MatchString(hostname)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
